tst/hashmap.tst

```diff
@@ -36,6 +36,35 @@ true
 gap> Set(List(KeyValueIterator(hashmap))) = List([1..1000],i->[i,i^2]);
 true
 
+# Check initial construction
+gap> Set(List(KeyValueIterator(HashMap([])))) = [];
+true
+gap> Set(List(KeyValueIterator(HashMap([[1,2],[3,4]])))) = [[1,2],[3,4]];
+true
+
+# printing
+gap> HashMap();
+HashMap([])
+gap> HashMap([]);
+HashMap([])
+gap> HashMap([[2,3]]);
+HashMap([[ 2, 3 ]])
+gap> String(HashMap([[2,3]]));
+"HashMap([[ 2, 3 ]])"
+gap> String(HashMap([[2,3],[3,4]])) in
+> ["HashMap([[ 2, 3 ], [ 3, 4 ]])","HashMap([[ 3, 4 ], [ 2, 3 ]])"];
```

Comment:
Off-topic, but sometimes I wonder if hash sets and maps should be made first class GAP features, and built into the kernel, and perhaps even receive some syntax support; like being able to write pairs as `A => B` and then you could do `HashMap( 1 => "a", 5 => "z" );` ...

Reply:
More importantly, if these were in the kernel and you could always rely on them, more people would know about them and use them etc.

Reply:
That sounds sensible. I wonder if we could generalise the notation for records, so users (or the kernel?) could easily put in their own function to call when creating record-like objects? There may be some reason this is a bad idea, so we could then introduce a new notation.

Reply:
I don't think that would work out great, at least not with what I proposed, where `a => b` would essentially evaluate to a pair expression (it could be a datatype of its own, or it could be just a list `[a, b]`).

Using the syntax `:=` for this this would then lead to ambiguity in the grammar because now `a := b` could be both an assignment statement, or an expression representing a pair.

Of course we could not lower `a => b` into a pair object  but instead give an expression of the form `IDENTIFIER( a := b, c := d, ...)` a special interpretation. But it's not clear to me how this should work exactly, and it seems rather specialized; to me it feels like we'd be painting ourselves into a corner here. Oh, and it would also potentially clash with another set of syntax extensions I'd love to see one day, namely keyword arguments / default values for arguments / allowing syntax calls with named arguments as in Python...


Anyway, right now the code for parsing record expressions hardcode the assumption that the keys are symbols; but presumably we'd want to allow arbitrary expressions, including function calls, so that one can do, say  `HashMap( f(1) => 3 + g("b"), "c" => 4 )`. But that means that no matter what, we wouldn't be able to reuse much if anything from the code for parsing record expressions either.

This all is why I would tend to introduce a new token to the scanner instead. Alas, maybe before doing any of that it'd be better to have hashsets / hashmaps in GAP itself, resp. ensure the ones here are more widely used?